be/src/olap/storage_engine.cpp

```diff
@@ -117,8 +119,179 @@ StorageEngine::~StorageEngine() {
+        clean_old_meta_func, OLD_HEADER_PREFIX);
+    if (clean_old_meta_status != OLAP_SUCCESS) {
+        // If failed to clean meta just skip the error, there will be useless metas in rocksdb column family
+        LOG(WARNING) << "there is failure when clean old tablet meta(olap header) from data dir:" << data_dir->path();
+    } else {
+        LOG(INFO) << "successfully clean old tablet meta(olap header) from data dir: " << data_dir->path();
+    }
+
+    // clean old files because they have hard links in new file name format
+    auto clean_old_files_func = [this, data_dir](long tablet_id,
+        long schema_hash, const std::string& value) -> bool {
+        TabletMetaPB tablet_meta_pb;
+        bool parsed = tablet_meta_pb.ParseFromString(value);
+        if (!parsed) {
+            // if errors when load, just skip it
+            LOG(WARNING) << "failed to load tablet meta from meta store to tablet=" << tablet_id << "." << schema_hash;
+            return true;
+        }
+
+        TabletSchema tablet_schema;
+        tablet_schema.init_from_pb(tablet_meta_pb.schema());
+        string data_path_prefix = data_dir->get_absolute_tablet_path(&tablet_meta_pb, true);
+
+        // convert visible pdelta file to rowsets and remove old files
+        for (auto& visible_rowset : tablet_meta_pb.rs_metas()) {
+            RowsetMetaSharedPtr alpha_rowset_meta(new AlphaRowsetMeta());
+            alpha_rowset_meta->init_from_pb(visible_rowset);
+            AlphaRowset rowset(&tablet_schema, data_path_prefix, data_dir, alpha_rowset_meta);
+            std::vector<std::string> old_files;
+            rowset.remove_old_files(&old_files);
+        }
+
+        // convert inc delta file to rowsets and remove old files
+        for (auto& inc_rowset : tablet_meta_pb.inc_rs_metas()) {
+            RowsetMetaSharedPtr alpha_rowset_meta(new AlphaRowsetMeta());
+            alpha_rowset_meta->init_from_pb(inc_rowset);
+            AlphaRowset rowset(&tablet_schema, data_path_prefix, data_dir, alpha_rowset_meta);
+            std::vector<std::string> old_files;
+            rowset.remove_old_files(&old_files);
+        }
+        return true;
+    };
+    OLAPStatus clean_old_tablet_status = TabletMetaManager::traverse_headers(data_dir->get_meta(), clean_old_files_func, HEADER_PREFIX);
+    if (clean_old_tablet_status != OLAP_SUCCESS) {
+        LOG(WARNING) << "there is failure when loading tablet and clean old files:" << data_dir->path();
+    } else {
+        LOG(INFO) << "load rowset from meta finished, data dir: " << data_dir->path();
+    }
+    return OLAP_SUCCESS;
+}
+
 // TODO(ygl): deal with rowsets and tablets when load failed
 OLAPStatus StorageEngine::_load_data_dir(DataDir* data_dir) {
+    // check if this is an old data path
+    bool is_tablet_convert_finished = false;
+    OLAPStatus res = data_dir->get_meta()->get_tablet_convert_finished(is_tablet_convert_finished);
```

Comment:
this flag is always true if the be upgrade from old version.Will it cause problems?

Reply:
This flag only indicates whether the old files have been converted or not. So it is always true if be upgrades from old version.